nntrainer/tensor/tensor_pool.cpp

```diff
@@ -0,0 +1,214 @@
+     * complete duration.
+     */
+    if (isTensorLongTerm(spec.lifespan)) {
+      validity_start = start_order;
+      validity_end = end_order;
+    }
+
+    /** 2. for each tensor request if it is in the provided range */
+    if (validity_end < start_order || validity_start > end_order)
+      continue;
+    validity_start = std::max(validity_start, start_order);
+    validity_end = std::max(validity_end, end_order);
+
+    /**
+     * 3. requestMemory for all the tensors and set their tokens
+     * @note +1 is to make the validity_end exlusive in the interval range
+     */
+    spec.token = mem_pool.requestMemory(spec.tensor->bytes(), validity_start,
+                                        validity_end + 1);
+    bytes_requested += spec.tensor->bytes();
+  }
+
+  /** 4. finalizeLayout for the memory pool. */
+  if (bytes_requested > 0)
+    mem_pool.planLayout(planner);
+}
+
+/**
+ * @brief Set the batch size for the inputs/outputs of the layers
+ */
+void TensorPool::setBatchSize(const std::string &name, unsigned int batch) {
+  if (pool.find(name) == pool.end())
+    throw std::invalid_argument("Requested tensor not found");
+
+  pool[name].tensor->updateBatch(batch);
+}
+
+/**
+ * @brief Allocate memory for all the managed tensors
+ */
+void TensorPool::allocate() {
+  mem_pool.allocate();
+
+  /** set the pointers using the token for all the tensors */
+  for (auto &entry : pool) {
+    auto &spec = entry.second;
+    spec.tensor->setData(mem_pool.getMemory(spec.token));
+  }
+}
+
+/**
+ * @brief Deallocate memory for all the managed tensors
+ */
+void TensorPool::deallocate() {
+  mem_pool.deallocate();
+
+  /** nullify the data pointers for the tensors */
+  for (auto &entry : pool) {
+    auto &spec = entry.second;
+    spec.tensor->setData(nullptr);
+  }
+}
+
+/**
+ * @brief     Expand the lifespan of the tensor with the given name
+ *
+ */
+void TensorPool::expand_lifespan(const std::string &name,
+                                 TensorLifespan lifespan) {
+  if (pool.find(name) == pool.end())
+    throw std::invalid_argument("Requested tensor not found");
+
+  auto &spec = pool[name];
+  spec.lifespan = enum_class_or<TensorLifespan>(spec.lifespan, lifespan);
+}
+
+/**
+ * @brief     Expand the execution order of the tensor with the given name
+ *
+ */
+void TensorPool::expand_lifespan(const std::string &name,
+                                 const std::vector<unsigned int> &exec_order) {
+  if (pool.find(name) == pool.end())
+    throw std::invalid_argument("Requested tensor not found");
+
+  auto &spec = pool[name];
+  spec.exec_order.insert(spec.exec_order.end(), exec_order.begin(),
+                         exec_order.end());
+}
+
+/**
+ * @brief     Check if the lifespan leads to long term valitidy
+ *
+ */
+bool TensorPool::isTensorLongTerm(const TensorLifespan &lifespan) {
+  switch (lifespan) {
+  case TensorLifespan::EPOCH_LIFESPAN:
+    /** fallthrough intended */
```

Comment:
it's just minor, but we can have compiler friendly [[fallthrough]] as of c++17 :)

Reply:
done.